Adds the technical definition of watchers to the specification

swcurran opened the pull request:
Adds the concept of watchers to the spec. As always, provides a relatively short/simple technical implementation, and leaves the governance / implementation details out of the spec.  That said, it does add more implementation details than, say, the witness implmentation.

Watchers are a list of URLs that monitor the DID. The URL must respond to a core set of 5 endpoints to return the cached DID Log, return the cached witness.json file, return a cached resource, webhook about an update to the DID, webhook about an update to a resource.

It's a slippery slope to define how much to include.

Questions from me:

1. Feedback on what is there so far?
2. What details need to be added about the list as DID Metadata?
3. Should the webhooks have an attached proof for "authorization"?




Signed-off-by: Stephen Curran <swcurran@gmail.com>

PatStLouis commented on spec/specification.md:
```diff
@@ -1151,6 +1156,76 @@ Implementer's Guide section on
+
+[[ref: Watchers]] need not be limited to those that collaborate with the [[ref: DID Controller]] and whose URLs are published in the [[ref: DID log]]. Anyone is free run a [[ref: watcher]] and monitor `did:webvh` DIDs. Such [[ref: watchers]] are not published in the [[ref: DID log]] and must rely on independent methods (e.g., polling) to track updates to the DID.
+
+##### Watcher Endpoints and Behaviour
+
+A [[ref: watcher]] that has its HTTP URL published in the [[ref: DID Log]] of a `did:webvh` must support the following endpoints. The query parameters **MUST NOT** be used in combination:
+
+- `HTTP GET <WATCHER URL>?vhscid=<SCID>` — Returns the most recent `did:webvh` [[ref: DID Log]] file for the DID with the provided `<SCID>` from the [[ref: watchers]] cache.
+
+- `HTTP GET <WATCHER URL>?vhscid-witness=<SCID>` — Returns the most recent `did:webvh` `witness.json` file for the DID with the provided `<SCID>` from the [[ref: watchers]] cache.
+
+- `HTTP GET <WATCHER URL>?vhscid=<SCID>&resource=<RESOURCE_PATH>` — Requests the cached version of the specified `<RESOURCE_PATH>` for the given `<SCID>`. If the resource is not found in the cache, the watcher **SHOULD** return a `404 Not Found` response.
+
+- `HTTP POST <WATCHER URL>?vhscid=<SCID>` — Sent by the [[ref: DID Controller]] or another authorized party to notify the [[ref: watcher]] that the `did:webvh` DID has been updated and should be retrieved from the DID's web location.
+
+- `HTTP POST <WATCHER URL>?vhscid-entry=<SCID>` — Sent by the [[ref: DID Controller]] or another authorized party to notify the [[ref: watcher]] that the `did:webvh` DID has been updated and provides, as the body of the message, the [[ref: JSON Line]] that is the new [[ref: DID log entry]].
+
+  - This endpoint can be used when the [[ref: watcher]] is also a [[ref: witness]] and is used to notify the [[ref: witness]] that a new entry has been created that the witness must verify and approve.
+
+- `HTTP POST <WATCHER URL>?vhscid=<SCID>&resource=<RESOURCE_PATH>` — Sent by the [[ref: DID Controller]] or another authorized party to notify the [[ref: watcher]] that a new [[ref: DID resource]] (file) has been published under the given `<RESOURCE_PATH>` and should be retrieved and cached.
```
Should we only have 1 endpoint defined or have a set of endpoints relevant to their actions? Having only 1 endpoint might overload this endpoint.

The operations a watcher should do can be summed up as:

- Return latest log entry for a given SCID
- Signal a new log entry is available for a given SCID
- Return witness file for a given SCID
- Request witness signature for a given log entry and SCID
  - Only on the case of a witness-watcher
- Return resource for a given SCID and resource path
- Signal new resource for a given SCID and resource path

I would suggest to divide these operations into 3 endpoints to avoid overloading:
`<WATCHER URL>/log`
`<WATCHER URL>/witness`
`<WATCHER URL>/resource`

Operations can be carried as such:

GET `<WATCHER URL>/log?scid=<SCID>`
Returns latest log entry.

POST `<WATCHER URL>/log?scid=<SCID>`
Empty body, fetch and update log entries.

GET `<WATCHER URL>/witness?scid=<SCID>`
Returns witness file.

POST `<WATCHER URL>/witness?scid=<SCID>`
```json
{
    "logEntry": {...},
    "options": {
        "callbackUrl": "https://..."
    }
}
```
Validates and signs the log entry provided in request body. This endpoint can return a 200 with the witness proof or a 202 with an empty body, having the proof returned as a POST request to the `callbackUrl` provided.

GET `<WATCHER URL>/resource?scid=<SCID>&path=<resourcePath>`
Returns resource

POST `<WATCHER URL>/resource?scid=<SCID>&path=<resourcePath>`
Fetch and updates resource

swcurran commented:
Updated the PR to add items discussed at the [2025-03-13 Meeting](https://hackmd.io/k4cIK9vQSlaeg2pdHE51IQ). Notably:

- Added a "DELETE" endpoint that requests removal of a SCID from the watcher's cache -- governance outside of this specification.
- Added that all POST and DELETE must respond to the successful receipt of a message with a 202, indicating asynchronous processing will occur.  This eliminates the possiblity of a POST of the entry for a witness responding in any way other than via the callback.
- Added that the URLs can be any type, and that if they are HTTP(S) URLs, they **MUST** adhere to the defined endpoints.

Not done yet:

- attach an OpenAPI definition of the API (@PatStLouis ?)
- define if/how any of the POSTs can/should be associated with the DID Controller -- a proof and if so the impact on the API. For example, should a DELETE have to have a proof, if it is possible that the DID Controller no longer exists?  Which is of course, back to governance.
- anything for the sender about POSTs/DELETE if a 202 is not received -- e.g. retries, etc.

brianorwhatever commented on spec/specification.md:
```diff
@@ -1156,6 +1161,71 @@ Implementer's Guide section on
+  - Retrieve a resource for a given [[ref: SCID]] and resource path.
+- **Notifications (typically) from the [[ref: DID Controller]]:**
+  - Notify the [[ref: watcher]] about a new log entry for a DID.
+  - Notify the [[ref: watcher]] about a new or updated [[ref: DID resource]].
```
I do have some concerns about replicating all of the resources for a DID with all of the watchers if this gets popular, and everyone is using this method for their resources, we are adding quite a big ask of these watchers..

swcurran commented on spec/specification.md:
```diff
@@ -1156,6 +1161,71 @@ Implementer's Guide section on
+[[ref: Watchers]] are components found in some digital trust and DID ecosystems that monitor DIDs on behalf of clients for various purposes, such as:
+
+- **Caching verified DIDs:** Storing verified DID documents to facilitate efficient resolution.
+- **Ensuring persistence:** Maintaining access to DIDs even after removal by the [[ref: DID Controller]]. For example, the [Verifiable Data Gateway](https://github.com/LedgerDomain/did-webplus?tab=readme-ov-file#verifiable-data-gateway-vdg) in [did](https://github.com/LedgerDomain/did-webplus/blob/main/README.md)[:webplus](https://github.com/LedgerDomain/did-webplus/blob/main/README.md) could fun ...
```
Reasonable.  My hope was that the it would encourage the did:webplus folks to formalize their VDG as a watcher as we have defined it.  It’s non-normative, so not a spec issue.

swcurran commented on spec/specification.md:
```diff
@@ -1156,6 +1161,71 @@ Implementer's Guide section on
+
+did:webvh provides two mechanisms for notifying resolvers (and their clients via [[spec:DID-RESOLUTION]] metadata) about configured [[ref: watchers]]:
+
+1. Witness-watchers: A [[ref: witness]] **MAY** also act as a [[ref: watcher]], indicated by the watcherURL attribute in the witnesses [[ref: parameter]]. If a [[ref: witness]] is removed, it ceases to be a [[ref: watcher]].
```
It can — it just has to be done explicitly. If the witness is dropped, they are removed from the witness and the (implied) witness-watcher list. If the DID Controller redefines the watcher list with that watcher, that’s fine.

brianorwhatever replied:
It feels like we have three different concepts now
1. a witness
2. a witness-watcher
3. a watcher

brianorwhatever replied:
Maybe 1 isn't then? This comes back to the question about how to get a proof from a witness. Previously, we had said this was out of spec and now we sort of have this in spec in the second category, but not in the first category.

swcurran replied:
Interesting idea…requiring that all witnesses be a watcher and provide a Watcher URI?  Others?

PatStLouis replied:
Strong -1 from me, as we have significant work done with the didcomm witness protocol and this would effectively scrap the current did webvh server work and acapy witnessing work.

PatStLouis replied:
It's not clear to me where the desire to have an endpoint to get a witness signature in the log file comes from. Who does this benefit? Surely the controller already knows how to get a witness signature prior to adding this witness to an entry. The witness parameter is there for resolvers to ensure the log has been properly witnessed.

swcurran replied:
We can leave it off — it was a bit of throw in and perhaps not well thought out.

It has value in that it defines an optional way to retrieve a witness proof.  A DID Controller/Witness is not forced to use it if the witness is also a watcher — it is just possible.  So the approach you are using is completely valid.  It is just that since we have this option, it can be used.  That said, we can remove it from the Watcher spec, and a DID Controller can have its witnesses use such as endpoint.

As I write this, I think it is best that we drop it from the spec — it is not obvious why it is there.   I’ll remove it, and let’s add an article to the information site about the different ways DID Controllers and Witnesses can interact — including what you have done, @PatStLouis and this approach.

brianorwhatever commented on spec/specification.md:
```diff
@@ -1027,7 +1030,8 @@ has the following data structure:
       {
-         "id": "<did:key DID of witness>"
+         "id": "<did:key DID of witness>",
+         "watcherURL": "<URL>"
```
I don't know if tying the two concepts together this tightly makes sense although I do understand the desire to have a URL for a witness. I don't think cheating and calling it a watcherURL is the way 😄

swcurran commented on spec/specification.md:
```diff
@@ -1156,6 +1161,71 @@ Implementer's Guide section on
+1. Witness-watchers: A [[ref: witness]] **MAY** also act as a [[ref: watcher]], indicated by the watcherURL attribute in the witnesses [[ref: parameter]]. If a [[ref: witness]] is removed, it ceases to be a [[ref: watcher]].
+2. Standalone watchers: The watchers [[ref: parameter]] lists URIs that identify independent [[ref: watchers]].
+
+Watcher URIs MAY use schemes other than HTTP(S), such as a DID ([[spec:DID-CORE]]), depending on the specific implementation or network requirements. However, this specification does not define how non-HTTP(S) [[ref: watcher]] URIs should be resolved or interacted with. If a [[ref: watcher]] uses an HTTP(S) URL, it **MUST** support the HTTP-based interaction model defined in the [Watcher Endpoint ...
```
That was guidance from the Work Item group (looking at @andrewwhitehead :smiling). The directive I received was to say that URIs were valid, but to provide API details only if the URI was an HTTP URL. Leaves the door open for someone defining (for example) a watcher with a DID URI and services for the Watcher endpoints.  But we definitely don’t want to put that into the spec.

PatStLouis commented on watcherOpenAPI/watcher-v1.0.0.yml:
```diff
@@ -0,0 +1,316 @@
+        '200':
+          description: successful operation
+          content:
+            text/jsonl:
```
[jsonl support for openapi is being discussed](https://github.com/OAI/OpenAPI-Specification/issues/3730), there's no clear way to define it at the moment besides documenting it.

swcurran replied:
We have this in the spec, `The content type for the did.jsonl file SHOULD be text/jsonl.`, so I think the OpenAPI spec should say that.  If not that, any suggestions?

PatStLouis replied:
yes I think that's good

PatStLouis replied:
I took some text from the spec and suggested a description for the response

swcurran commented:
@PatStLouis -- removed the witness entry request from the API. I think I got the OpenAPI right.  Please review.

PatStLouis commented on watcherOpenAPI/watcher-v1.0.0.yml:
```diff
@@ -0,0 +1,241 @@
+            domain:
+              type: string
+              example: 'https://watcher.example.com'
+    WitnessProof:
+      type: object
+      properties:
+        versionId:
+          type: string
+        proof:
+          $ref: '#/components/schemas/DataIntegrityProof'
+    WitnessFile:
+      type: array
+      items:
+        $ref: '#/components/schemas/WitnessProof'
```
```suggestion
    WitnessFile:
      type: object
      properties:
        versionId:
          type: string
        proof:
          type: array
          items:
            $ref: '#/components/schemas/DataIntegrityProof'
```

PatStLouis commented on watcherOpenAPI/watcher-v1.0.0.yml:
```diff
@@ -0,0 +1,241 @@
+      items:
+        $ref: '#/components/schemas/WitnessProof'
+    LogFile:
+      type: string
```
```suggestion
      type: string
      description: The DID log file contains a list of entries, one for each version of the DID. Each entry has (per the JSON Lines specification) all extra whitespace removed, a \n character appended, and the result added to the DID Log.
```

PatStLouis commented on watcherOpenAPI/watcher-v1.0.0.yml:
```diff
@@ -0,0 +1,241 @@
+    LogFile:
+      type: string
+      example: |
+          {"versionId": "", "versionTime": "", "parameters": {}, "state": {}, "proof": {}}
+          {"versionId": "", "versionTime": "", "parameters": {}, "state": {}, "proof": {}}
+          {"versionId": "", "versionTime": "", "parameters": {}, "state": {}, "proof": {}}
```
```suggestion
          {"versionId":"","versionTime":"","parameters":{},"state":{},"proof":[]}
          {"versionId":"","versionTime":"","parameters":{},"state":{},"proof":[]}
          {"versionId":"","versionTime":"","parameters":{},"state":{},"proof":[]}
```

PatStLouis reviewed (changes requested):
Some small tweaks
